### PR TITLE
Latest features update

### PR DESCRIPTION
**Pull request by stelladk**
- Mini-batch gradient descent
- Bayesian information criterion
- GPU tracking with codecarbon

### CODE REVIEW

**Comment by stelladk** [comment]
Will need more tests!

File: src/gromo/growing_module.py
Comment (TheoRudkiewicz) [inline src/gromo/growing_module.py:1]: I think that batch normalization could already be done with a `post_addition_function` that would include batch normalization.  If so I think it is preferable to keep the core implementation parsimonious.
Otherwise, I would advocate for taking as parameters directly the batch normalization function (like we did with the activation function) instead of a boolean. In addition, if we implement it for the Addition module we should add it also to the GrowingModule.

Comment (stelladk) [reply src/gromo/growing_module.py:1]: Okay I will work on that

Comment (TheoRudkiewicz) [reply src/gromo/growing_module.py:1]: You don't like the idea of using directly the `post_addition_function` for both activation and normalization ? Maybe we can discuss this IRL.

File: src/gromo/growing_module.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ class AdditionGrowingModule(torch.nn.Module):
     def __init__(
         self,
         post_addition_function: torch.nn.Module = torch.nn.Identity(),
+        post_addition_normalization: torch.nn.Module = torch.nn.Identity(),
```
Comment (sylvchev) [inline src/gromo/growing_module.py:20]:
```suggestion
```

File: src/gromo/growing_module.py [hunk 2]
```diff
@@ -40,6 +41,9 @@ def __init__(
         self.post_addition_function: torch.nn.Module = post_addition_function
         if self.post_addition_function:
             self.post_addition_function = self.post_addition_function.to(self.device)
+        self.post_addition_normalization: torch.nn.Module = post_addition_normalization
+        if self.post_addition_normalization:
+            self.post_addition_normalization = self.post_addition_normalization.to(device)
```
Comment (sylvchev) [inline src/gromo/growing_module.py:46]:
```suggestion
```

File: src/gromo/growing_module.py [hunk 3]
```diff
@@ -174,8 +178,12 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
             self.input.retain_grad()
 
-        if (self.post_addition_function) and (x is not None):
-            y = self.post_addition_function(x)
+        if x is not None:
+            if self.post_addition_normalization:
+                x = self.post_addition_normalization(x)
+
+            if self.post_addition_function:
+                y = self.post_addition_function(x)
```
Comment (sylvchev) [inline src/gromo/growing_module.py:186]:
```suggestion
        if (self.post_addition_function) and (x is not None):
            y = self.post_addition_function(x)
```

File: src/gromo/linear_growing_module.py [hunk 4]
```diff
@@ -12,6 +12,7 @@ class LinearAdditionGrowingModule(AdditionGrowingModule):
     def __init__(
         self,
         post_addition_function: torch.nn.Module = torch.nn.Identity(),
+        post_addition_normalization: torch.nn.Module = torch.nn.Identity(),
```
Comment (sylvchev) [inline src/gromo/linear_growing_module.py:15]:
```suggestion
 ```

File: src/gromo/linear_growing_module.py [hunk 5]
```diff
@@ -27,6 +28,7 @@ def __init__(
         # TODO: check if we can automatically get the input shape
         super(LinearAdditionGrowingModule, self).__init__(
             post_addition_function=post_addition_function,
+            post_addition_normalization=post_addition_normalization,
```
Comment (sylvchev) [inline src/gromo/linear_growing_module.py:31]:
```suggestion
```

File: src/gromo/graph_network/GrowableDAG.py [hunk 6]
```diff
@@ -287,14 +314,22 @@ def update_nodes(
...
                 self.__set_node_module(
                     node,
                     LinearAdditionGrowingModule(
                         allow_growing=True,
-                        in_features=self.nodes[node]["size"],
+                        in_features=in_features,
                         post_addition_function=activation_fn(
                             self.nodes[node].get("activation")
                         ),
+                        post_addition_normalization=batch_norm,
```
Comment (sylvchev) [inline src/gromo/graph_network/GrowableDAG.py:332]:
```suggestion
                        in_features=in_features,
                        post_addition_function=torch.nn.Sequential(
                            batch_norm,
                            activation_fn(self.nodes[node].get("activation"),
                        ),
```

**Review by sylvchev — COMMENTED** [review]
For the batch norm, I could be better to use the post_addition_function making it a composition of the addition and the normalization. THis will allows to keep the same code for linear and conv modules.
Here are code suggestion to revert to post_addition_function alone and an example of composition in the growableDAG code.

File: src/gromo/utils/gpu_tracking.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        self.gpu_index = gpu_index
+        self.interval = interval
+        self._logger = logger
+        # self.tracking = False
+        # self.thread = None
+        self._tracker = OfflineEmissionsTracker(
+            gpu_ids=gpu_index,
+            measure_power_secs=interval,
+            allow_multiple_runs=True,
+            country_iso_code="FRA",
```
Comment (TheoRudkiewicz) [inline src/gromo/utils/gpu_tracking.py:35]: Maybe add `country_iso_code` to the parameters of the __init__.

File: src/gromo/utils/utils.py [hunk 8]
```diff
@@ -170,6 +170,137 @@ def line_search(cost_fn: Callable, verbose: bool = True) -> tuple[float, float]:
...
+            )
+    else:
+        parameters = model.parameters()
+        saved_parameters = list(model.parameters())
+    optimizer = torch.optim.AdamW(parameters, lr=lrate, weight_decay=0)
+
+    for epoch in range(max_epochs):
+        correct, total, epoch_loss = 0, 0, 0
+        for x_batch, y_batch in dataloader:
+            optimizer.zero_grad(set_to_none=True)
```
Comment (TheoRudkiewicz) [inline src/gromo/utils/utils.py:239]: If I understand it correctly you added the `set_to_none` option when you were searching for the bug fixed with safe_forward. This is probably not useful anymore ? If it is unnecessary removing it will porbably make the function faster as it avoid re-affecting memory each time.

File: tests/test_dag_growing_network.py [hunk 9]
```diff
@@ -58,7 +63,7 @@ def test_init_empty_graph(self) -> None:
         assert self.net.dag.nodes["end"]["size"] == self.out_features
         assert self.net.dag.nodes["start"]["type"] == "L"
         assert self.net.dag.nodes["end"]["type"] == "L"
-        # assert self.net.dag.nodes["end"]["use_batch_norm"] == False
+        assert self.net.dag.nodes["end"]["use_batch_norm"] == False
```
Comment (TheoRudkiewicz) [inline tests/test_dag_growing_network.py:66]:
```suggestion
        assert self.net.dag.nodes["end"]["use_batch_norm"] is False
```
i think that you are trying to test that the value is *exactly* False. This is slightly different: 
```
0 is False 
>>> False
0 == False 
>>> True
False is False
>>> True
```

**Review by TheoRudkiewicz — COMMENTED** [review]
Did not went into full details, seems good overall.

File: src/gromo/graph_network/GrowableDAG.py [hunk 10]
```diff
@@ -30,6 +30,30 @@
...
+    ), f"Input shape {input.shape} must match the input feature size. Expected: {self.in_features}, Found: {input.shape[1]}"
+    if self.in_features == 0:
+        return torch.zeros(
+            input.shape[0], self.out_features, device=global_device(), requires_grad=True
+        )
+    return nn.functional.linear(input, self.weight, self.bias)
+
+
 class GrowableDAG(nx.DiGraph, nn.Module):
     def __init__(
```
[the remark this answers is not in the document]
Comment (stelladk) [reply src/gromo/graph_network/GrowableDAG.py:58]: This does not create a problem here because the dictionary is never modified. If you think it is a deal-breaker we can change it

File: src/gromo/utils/gpu_tracking.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    interval : int, optional
+        time interval between measuring, by default 15
+    logger : Logger | None, optional
+        associated logger to track metrics, by default None
+    """
+
+    def __init__(
+        self, gpu_index: list[int] = [0], interval: int = 15, logger: Logger | None = None
+    ) -> None:
+        self.gpu_index = gpu_index
```
[the remark this answers is not in the document]
Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:26]: Same here, the gpu_index is not modified

File: src/gromo/graph_network/GrowableDAG.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/graph_network/GrowableDAG.py:58]: No its ok if your confident it won't lead to bugs.

File: src/gromo/utils/gpu_tracking.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
+from types import TracebackType
+from typing import Optional, Self, Type
+
+import numpy as np
+from codecarbon import OfflineEmissionsTracker
```
[the remark this answers is not in the document]
Comment (sylvchev) [reply src/gromo/utils/gpu_tracking.py:5]: We could add codecarbon as an optional dependency. Right now, we already have two optional sets in pyproject.toml: test and docs, with dependencies related to unit tests and documentation building. 

Is this a hard dependency or an optional one? I clearly prefer if this dependency is optional, as there is a high risk of bugs or dependencies blocks with hardware-related library. If it is optional, we could add another optional set in project.toml, called `energy` or `carbon`, that includes only codecarbon.

File: src/gromo/utils/gpu_tracking.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        the country of the cluster in iso code to estimate carbon emissions
+    logger : Logger | None, optional
+        associated logger to track metrics, by default None
+    """
+
+    def __init__(
+        self,
+        gpu_index: list[int] = [0],
+        interval: int = 15,
+        country_iso_code: str = "FRA",
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:29]: I think this should be set in global settings, but this is outside of the scope of this PR.

Could you open an issue,  or add this to the issue for the global settings, to remember that we need to take care of it?

For the global settings, we could inspire from codecarbon code: https://mlco2.github.io/codecarbon/usage.html#configuration

File: src/gromo/utils/gpu_tracking.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        if self._logger is not None:
+            for key, value in self.gpu_metrics.items():
+                if key in ["timestamp", "project_name", "run_id", "experiment_id"]:
+                    continue
+                if not isinstance(value, (int, float)):
+                    self._logger.log_parameter(key, str(value))
+                else:
+                    self._logger.save_metric(name=f"codecarbon/{key}", value=value)
+
+    # @property
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:73]: Do we still need those comment? You could open an issue if you want to add further code later.

**Comment by sylvchev** [comment]
Could you convert your assert into self.assert* in the tests?

File: src/gromo/utils/gpu_tracking.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:5]: Only this specific file uses codecarbon and it is used as a context manager in a script. It is fully optional. So far I have not implemented any tests on gpu_tracking because it is quite simple and thus we avoid issues with the CI. We might choose to change the energy tracking library any time.

File: src/gromo/utils/gpu_tracking.py (hunk 13 above)
Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:29]: Opened at #30 !

File: src/gromo/graph_network/GrowableDAG.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sylvchev) [reply src/gromo/graph_network/GrowableDAG.py:58]: @stelladk You could mark the conversation as resolved

File: src/gromo/utils/gpu_tracking.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sylvchev) [reply src/gromo/utils/gpu_tracking.py:5]: In that case, I suggest we add an optional set (like docs and test) called `carbon`, that contains only codecarbon for now and that we set up a conditional import for codacarbon here (raising an error if it is not installed).

File: src/gromo/utils/gpu_tracking.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sylvchev) [reply src/gromo/utils/gpu_tracking.py]: you could resolve here as well @stelladk

File: src/gromo/utils/gpu_tracking.py (hunk 13 above)
Comment (sylvchev) [reply src/gromo/utils/gpu_tracking.py:41]: thanks

**Review by sylvchev — COMMENTED** [review]
Could you add codecarbon in optional set in pyproject.toml and do a conditional import in gpu_tracking.py?
We could merge right after.

File: src/gromo/utils/gpu_tracking.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(
+        self,
+        gpu_index: list[int] = [0],
+        interval: int = 15,
+        country_iso_code: str = "FRA",
+        logger: Logger | None = None,
+    ) -> None:
+        self.gpu_index = gpu_index
+        self.interval = interval
+        self._logger = logger
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:46]:
```suggestion
        self.logger = logger
```

The private attributes are written with leading _
Logger is a public attribute, any reason to make it private?

File: src/gromo/utils/gpu_tracking.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+                    if key in ["timestamp", "project_name", "run_id", "experiment_id"]:
+                        continue
+                    if not isinstance(value, (int, float)):
+                        self._logger.log_parameter(key, str(value))
+                    else:
+                        self._logger.save_metric(name=f"codecarbon/{key}", value=value)
+
+    def __import_module(self) -> None:
+        try:
+            global codecarbon
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:92]:
```suggestion
```
Better to avoid global

File: src/gromo/utils/gpu_tracking.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+            import codecarbon
+
+            self.tracking = True
+        except ImportError as err:
+            warnings.warn(f"{err}. Energy tracking will be skipped.", ImportWarning)
+            self.tracking = False
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:98]:
```suggestion
            tracking = False

        return tracking
```

File: src/gromo/utils/gpu_tracking.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        interval: int = 15,
+        country_iso_code: str = "FRA",
+        logger: Logger | None = None,
+    ) -> None:
+        self.gpu_index = gpu_index
+        self.interval = interval
+        self._logger = logger
+        # self.thread = None
+
+        self.__import_module()
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:49]:
```suggestion
        self.tracking = self.__import_module()
```
It is better to expose all parameters in init for code maintainability. Returning the value of tracking makes it explicit where it is defined.

File: src/gromo/utils/gpu_tracking.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+            if self._logger is not None:
+                for key, value in self.gpu_metrics.items():
+                    if key in ["timestamp", "project_name", "run_id", "experiment_id"]:
+                        continue
+                    if not isinstance(value, (int, float)):
+                        self._logger.log_parameter(key, str(value))
+                    else:
+                        self._logger.save_metric(name=f"codecarbon/{key}", value=value)
+
+    def __import_module(self) -> None:
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:90]:
```suggestion
    def __import_module(self) -> bool:
```

File: src/gromo/utils/gpu_tracking.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        try:
+            global codecarbon
+            import codecarbon
+
+            self.tracking = True
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:95]:
```suggestion
            tracking = True
```

File: src/gromo/utils/gpu_tracking.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        """Stop tracking power usage when exiting the context."""
+        if self.tracking:
+            self._tracker.stop()
+            self.gpu_metrics = self._tracker.final_emissions_data.__dict__
+            self.gpu_metrics["total_power"] = np.sum(
+                [hardware.total_power().kW * 1000 for hardware in self._tracker._hardware]
+            )
+            if self._logger is not None:
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:81]:
```suggestion
            if self.logger is not None:
```

File: src/gromo/utils/gpu_tracking.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+                for key, value in self.gpu_metrics.items():
+                    if key in ["timestamp", "project_name", "run_id", "experiment_id"]:
+                        continue
+                    if not isinstance(value, (int, float)):
+                        self._logger.log_parameter(key, str(value))
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:86]:
```suggestion
                        self.logger.log_parameter(key, str(value))
```

File: src/gromo/utils/gpu_tracking.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+                    else:
+                        self._logger.save_metric(name=f"codecarbon/{key}", value=value)
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:88]:
```suggestion
                        self.logger.save_metric(name=f"codecarbon/{key}", value=value)
```

File: src/gromo/utils/gpu_tracking.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    gpu_index : list[int], optional
+        indices of gpus to track, by default [0]
+    interval : int, optional
+        time interval between measuring, by default 15
+    country_iso_code : str, optional
+        the country of the cluster in iso code to estimate carbon emissions
+    logger : Logger | None, optional
+        associated logger to track metrics, by default None
+
+
```
Comment (sylvchev) [inline src/gromo/utils/gpu_tracking.py:24]:
```suggestion
    Attributes
    -----------------
    tracking: bool
        Define if tracker is activated

    _tracker: Tracker
        Tracker used to monitor GPU

    gpu_metrics: dict
        Metrics returned by logger
```

**Review by sylvchev — APPROVED** [review]
Some last corrections, you could integrate and tests those comments and then we could merge!

File: src/gromo/utils/gpu_tracking.py (hunk 15 above)
Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:46]: In order not to be accessed through the GPU tracker

File: src/gromo/utils/gpu_tracking.py (hunk 16 above)
Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:92]: We cannot dynamically import the library without it

Comment (stelladk) [reply src/gromo/utils/gpu_tracking.py:92]: Another options is:
```
moduleName = "codecarbon"
globals()[moduleName] = import_module(moduleName)
```
